src/dune_pkg/opam_solver.ml

```diff
@@ -132,27 +132,21 @@ module Context_for_dune = struct
   ;;
 
   let filter_deps t package filtered_formula =
-    let solver_env =
-      let package_is_local =
-        OpamPackage.Name.Map.mem (OpamPackage.name package) t.local_packages
-      in
-      if package_is_local
-      then t.solver_env
-      else
+    let solver_env, test, doc =
+      match OpamPackage.Name.Map.mem (OpamPackage.name package) t.local_packages with
+      | false ->
         (* Flag variables pertain only to local packages. This is because these
            variables enable dependencies on test and documentation packages and
            we don't want to pull in test and doc dependencies for dependencies
            of local packages. *)
-        Solver_env.clear_flags t.solver_env
+        Solver_env.clear_flags t.solver_env, false, false
+      | true ->
+        let test = Solver_env.flag t.solver_env `With_test in
+        let doc = Solver_env.flag t.solver_env `With_doc in
+        t.solver_env, test, doc
     in
     map_filters filtered_formula ~f:(resolve_solver_env solver_env)
-    |> OpamFilter.filter_deps
-         ~build:true
-         ~post:false
-         ~dev:false
-         ~default:false
-         ~test:false
-         ~doc:false
+    |> OpamFilter.filter_deps ~build:true ~post:false ~dev:false ~default:false ~test ~doc
```

Comment:
I'm surprised it's necessary to explicitly pass `~test` and `~doc` here. The `resolve_solver_env` function should have already substituted all occurrences of `with-test` and `with-doc` with `true` or `false` so those variables shouldn't be present in the filters passed to `filter_deps` here. Does the behaviour change if you leave `~test` and `~doc` as `false`?

Reply:
You're right. It doesn't do anything. I wonder why we need the `filter_deps` step at all then. We could include build, dev, etc all in the Solver_env.t

Reply:
I changed the PR to just include the tests.